klemens_basic_thoughts_1.md

```diff
@@ -0,0 +1,206 @@
+# Design Proposal
+
+# Concepts
+
+For extensibility, we have two core concepts
+
+## ProcessLauncher
+
+The process Launcher is the class that starts the actual process: 
+
+```cpp
+template<typename T>
+concept bool ProcessLauncher() {
+    return requires(T launcher) {
+        {launcher.set_error(std::error_code())} -> void; //so initializers can post internal errors
+        {launcher()} -> process; //refine that so check tha parameter list, good enough for now.
+    };
+}
+```
+
+The reason to make this a concept is so replacement is possible. The use case in mind is `vfork` -> vfork is built into the posix-executor (renamed to launcher here) in boost-process and can be activated by a parameter. This is ugly and should not be in the standard; yet vfork as explicitly requested, and the possibility I see is to allow users to provide their own launcher OR the vendor to provide a `std::ext::vfork_launcher`. 
+
+Having this customization point does however allow more things, like security or virtualization.
+
+## ProcessInitializer
+
+The process initializer is a class that does some work to set up a process.
+
+```cpp
+template<typename Init, typename Launcher = process_launcher>
+concept bool ProcessInitializer = requires(Init initializer, Launcher launcher) {
+        {initializer.on_setup(launcher)}   -> void;
+        {initializer.on_success(launcher)} -> void;
+        {initializer.on_error(launcher, std::error_code())} -> void;
+    };
+```
+
+A vendor implementation may provide more customization points, such as handling fork etc. Note however, that it is required to invoke the three given one from the launching process.
+
+## ProcessStream
+
+Anything that contains native_handle (HANDLE* on windows & int on linux) that can be used as a stream. Includes TCP Ports btw.
+
+# Launching a process
+
+## Core function
+
+Launching a process is done by default through the destructor of `std::process`. There are two modes to launch a process in theory: providing a `path` & arguments, OR to provide a command string. The latter will be interpreted by a shell (cmd.exe or /bin/sh). This is considered unsafe (because shell injection), and `std::system` does exactly that at the moment. Point of discussion: does that make sense to provide? Or should be move to deprecate `std::system` since we then would have a process library?
```

Comment:
I assume you mean 'constructor'?

Reply:
Hah, yep. Launching a process in a destructor is probably possible, but... a questionable design choice.